### PR TITLE
 feat: add meta property and rollupVersion to plugin context

### LINKED ISSUE #1374: Plugins need access to rollup version and/or to plugin API version
Some plugin API changes break plugins when lower than expected rollup version is used. 

For example warn()/error() on context given to transform() and any new entry point ever added.

A sane way of retrieving rollup version would help a lot. Having a plugin API version would be ideal.

### PR DESCRIPTION
**Pull request by shellscape**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [x] tests
- [ ] documentation
- [ ] metadata

### Breaking Changes?

- [ ] yes
- [x] no

If yes, please describe the breakage.

### Please Describe Your Changes

This resolves #1374 ~~by changing the signature of the `options` hook to accept a second parameter named `meta` which currently will contain `{ version: string }`.~~ by adding a `meta` property to a plugin context (`this` within hooks), which contains `rollupVersion` as suggested.

I ended up piggy backing on the test ~~that checks the `options` hook can effect `inputOptions`. If there's a better place for that, I'm all for it.~~ which checks validity of `buildStart` and `buildEnd` hooks. The filename is just plain old `hooks`, so it seemed an apt place for them. (This is probably a case of the tests needing a bit of spring cleaning.)

### CODE REVIEW

**Comment by shellscape** [comment]
I can definitely take a look at going that route. I'm a big fan of the `meta` property name to contain non-essential metadata. Any objection to that naming?

**Comment by guybedford** [comment]
I'd be happy with that.

**Comment by shellscape** [comment]
@guybedford changes made, tests passed locally. ESLint was responsible for the whitespace changes in `hooks/index.js` and they appear legitimate changes.

File: test/function/samples/plugins-can-manipulate-options/_config.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ module.exports = {
 	options: {
 		plugins: [
 			{
-				options(options) {
+				options(options, meta) {
```
Comment (guybedford) [inline test/function/samples/plugins-can-manipulate-options/_config.js]: This should no longer be needed.

**Comment by shellscape** [comment]
Lot's of opportunity for hypotheticals, might include:

- metrics of varying kinds (plugin run times, cache metrics, ...)
- plugin info we could populate automatically, which could be used in future custom build reports/reporters

Mostly I've used `meta` to future-proof as a home for static bits of information, so as not to pollute (used in loose terms) a namespace or object properties. I'm not held strongly to using that and wouldn't push back on tacking `rollupVersion` straight onto the context, if `meta` doesn't sit right.

**Comment by guybedford** [comment]
> metrics of varying kinds (plugin run times, cache metrics, ...)

Currently we have a `getTimings` method provided on the bundle result.

Not at all debating - just trying to think about what info might be provided so that we can make sure the convention can scale well.

**Comment by shellscape** [comment]
No worries. We can let this marinate for a bit too.

**Comment by lukastaegert** [comment]
Very nice, this could really help in migrating plugins to new APIs by allowing them to support an old API in parallel. Some thoughts and questions:

* How do we want to decide in the future if stuff is put onto `meta` or directly onto the plugin context? For now, one possible distinction could be that `meta` is for static information while the context is for API functions?
* In this spirit as a hypothetical example, one way to make timings available to plugins here could be: If the `perf` option is provided, `meta` contains an additional property that contains the currently collected timing information and which is mutated each time new timings are added.
* How do we want to handle documentation? This should be documented soon, otherwise we might forget about it. This means we should be sure enough about this because people will start using it. For reference, this is where this information is collected at the moment: https://github.com/rollup/rollup/wiki/Plugins#plugin-context

**Comment by tivac** [comment]
I really hate that the plugin API lives in the wiki, because I always forget that the wiki even exists now that it's been 99% deprecated.

**Comment by shellscape** [comment]
@tivac I'm absolutely planning to merge that to the main docs repo this week.

@lukastaegert I agree that it should only contain static data. That's always my intention when I use that moniker. And the ideas around perf data seem reasonable to me. I'll also update the wiki before this is merged. (and following, will be a PR on the docs site to move the plugin info)

Can we get a second approval to knock this one home?

**Comment by shellscape** [comment]
Thanks @lukastaegert I'll work on the documentation later tonight or tomorrow. Keep an eye out for that one.

**Pull request merged**